libevmasm/SimplificationRules.cpp

```diff
@@ -163,6 +163,14 @@ Rules::Rules()
 		{{Instruction::AND, {{{X}, {Instruction::OR, {X, Y}}}}}, [=]{ return X; }},
 		{{Instruction::AND, {{{X}, {Instruction::NOT, {X}}}}}, [=]{ return u256(0); }},
 		{{Instruction::OR, {{{X}, {Instruction::NOT, {X}}}}}, [=]{ return ~u256(0); }},
+
+		{{Instruction::MOD, {X, A}}, [=]() -> Pattern {
+			// if power of 2, transform to AND
+			// this optimisation makes sense as long as the gas cost of AND is less than MOD
+			if (A.d() > 0 && (A.d() & (A.d() - 1)) == 0)
+				return {Instruction::AND, {X, A.d() - 1}};
+			return {Instruction::MOD, {X, A}};
```

Comment:
This will cause a recursion and a crash with out of stack.

@chriseth is there a simple way in the pattern system to avoid this? If not, it can be implemented in peephole optimiser.

Reply:
Ah right, I didn't think of that. Another solution would be to add one rule for all 256 powers of two.

Reply:
The rules are indexed by opcode, so it is probably not too bad for performance.

Reply:
Yeah probably that's the best alternative. In that case we need only need to ensure pattern matching is optimised :)